File: android/library/src/main/java/ai/mlc/mlcllm/ChatModule.java
package ai.mlc.mlcllm;

import org.apache.tvm.Device;
import org.apache.tvm.Function;
import org.apache.tvm.Module;

public class ChatModule {
    private Function reloadFunc;
    private Function unloadFunc;
    private Function prefillFunc;
    private Function decodeFunc;
    private Function getMessage;
    private Function stoppedFunc;
    private Function resetChatFunc;
    private Function runtimeStatsTextFunc;
    private Module llmChat;

    public ChatModule() {
        Function createFunc = Function.getFunction("mlc.llm_chat_create");
        assert createFunc != null;
        llmChat = createFunc.pushArg(Device.opencl().deviceType).pushArg(0).invoke().asModule();
        reloadFunc = llmChat.getFunction("reload");
        unloadFunc = llmChat.getFunction("unload");
        prefillFunc = llmChat.getFunction("prefill");
        decodeFunc = llmChat.getFunction("decode");
        getMessage = llmChat.getFunction("get_message");
        stoppedFunc = llmChat.getFunction("stopped");
        resetChatFunc = llmChat.getFunction("reset_chat");
        runtimeStatsTextFunc = llmChat.getFunction("runtime_stats_text");
    }

    public void unload() {
        unloadFunc.invoke();
    }

    public void reload(String modelLib, String modelPath) {
        String libPrefix = modelLib.replace('-', '_') + "_";
        Function systemLibFunc = Function.getFunction("runtime.SystemLib");
        assert systemLibFunc != null;
        systemLibFunc = systemLibFunc.pushArg(libPrefix);
        Module lib = systemLibFunc.invoke().asModule();
        reloadFunc = reloadFunc.pushArg(lib).pushArg(modelPath);
        reloadFunc.invoke();
    }

    public void resetChat() {
        resetChatFunc.invoke();
    }

    public void prefill(String input) {
        prefillFunc.pushArg(input).invoke();
    }

    public String getMessage() {
        return getMessage.invoke().asString();
    }

    public String runtimeStatsText() {
        return runtimeStatsTextFunc.invoke().asString();
    }

    public void evaluate() {
        llmChat.getFunction("evaluate").invoke();
    }

    public boolean stopped() {
        return stoppedFunc.invoke().asLong() != 0L;
    }

    public void decode() {
        decodeFunc.invoke();
    }
}

File: android/MLCChat/app/src/main/java/ai/mlc/mlcchat/ChatModule.kt
package ai.mlc.mlcchat

import android.os.Looper
import android.util.Log
import org.apache.tvm.Device
import org.apache.tvm.Function
import org.apache.tvm.Module

class ChatModule {
    private var reloadFunc: Function
    private var unloadFunc: Function
    private var prefillFunc: Function
    private var decodeFunc: Function
    private var getMessage: Function
    private var stoppedFunc: Function
    private var resetChatFunc: Function
    private var runtimeStatsTextFunc: Function
    private var llmChat: Module

    init {
        val fcreate = Function.getFunction("mlc.llm_chat_create")
        require(fcreate != null)
        llmChat = fcreate.pushArg(Device.opencl().deviceType).pushArg(0).invoke().asModule()

        reloadFunc = llmChat.getFunction("reload")
        unloadFunc = llmChat.getFunction("unload")
        prefillFunc = llmChat.getFunction("prefill")
        decodeFunc = llmChat.getFunction("decode")
        getMessage = llmChat.getFunction("get_message")
        stoppedFunc = llmChat.getFunction("stopped")
        resetChatFunc = llmChat.getFunction("reset_chat")
        runtimeStatsTextFunc = llmChat.getFunction("runtime_stats_text")
    }

    fun unload() {
        require(!Looper.getMainLooper().isCurrentThread)
        unloadFunc.invoke()
    }

    fun reload(modelLib: String, modelPath: String) {
        require(!Looper.getMainLooper().isCurrentThread)
        var libPrefix = modelLib
        Log.i(TAG, "lib_prefix: $libPrefix")
        libPrefix = libPrefix.replace('-', '_')
        libPrefix += "_"
        Log.i(TAG, "lib_prefix: $libPrefix")
        var systemLibFunc = Function.getFunction("runtime.SystemLib")
        require(systemLibFunc != null)
        Log.i(TAG, "system_lib_func: $systemLibFunc")
        systemLibFunc = systemLibFunc.pushArg(libPrefix)
        val lib = systemLibFunc.invoke().asModule()
        require(lib != null)
        Log.i(TAG, "lib: $lib")
        Log.i(TAG, "modelPath: $modelPath")
        Log.i(TAG, "reload_func: $reloadFunc")
        reloadFunc = reloadFunc.pushArg(lib).pushArg(modelPath)
        reloadFunc.invoke()
    }

    fun resetChat() {
        require(!Looper.getMainLooper().isCurrentThread)
        resetChatFunc.invoke();
    }

    fun prefill(input: String) {
        require(!Looper.getMainLooper().isCurrentThread)
        prefillFunc.pushArg(input).invoke();
    }

    fun getMessage(): String {
        require(!Looper.getMainLooper().isCurrentThread)
        return getMessage.invoke().asString()
    }

    fun runtimeStatsText(): String {
        require(!Looper.getMainLooper().isCurrentThread)
        return runtimeStatsTextFunc.invoke().asString()
    }

    fun evaluate() {
        llmChat.getFunction("evaluate").invoke()
    }

    fun stopped(): Boolean {
        require(!Looper.getMainLooper().isCurrentThread)
        return stoppedFunc.invoke().asLong() != 0L
    }

    fun decode() {
        require(!Looper.getMainLooper().isCurrentThread)
        decodeFunc.invoke()
    }

    companion object {
        private const val TAG = "ChatModule"
    }
}